crates/subgraph/src/performance/order_performance.rs

```diff
@@ -0,0 +1,765 @@
+        acc += new_val.0;
+    } else if sign {
+        if new_val.0 >= acc {
+            sign = false;
+            acc = new_val.0 - acc;
+        } else {
+            acc -= new_val.0;
+        }
+    } else if acc >= new_val.0 {
+        acc -= new_val.0;
+    } else {
+        sign = true;
+        acc = new_val.0 - acc;
+    }
+    (acc, sign)
+}
+
+/// helper struct that provides sorting tokens based on a given net vol
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct TokenBasedVol<'a> {
+    token: &'a Erc20,
+    net_vol: U256,
+    is_neg: bool,
+}
+impl<'a> Ord for TokenBasedVol<'a> {
+    fn clamp(self, _min: Self, _max: Self) -> Self
+    where
+        Self: Sized,
+        Self: PartialOrd,
+    {
+        self
+    }
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.is_neg == other.is_neg {
+            match self.net_vol.cmp(&other.net_vol) {
+                Ordering::Greater => {
+                    if self.is_neg {
+                        Ordering::Less
+                    } else {
+                        Ordering::Greater
+                    }
+                }
+                Ordering::Less => {
+                    if self.is_neg {
+                        Ordering::Greater
+                    } else {
+                        Ordering::Less
+                    }
+                }
+                Ordering::Equal => Ordering::Equal,
+            }
+        } else if self.is_neg {
+            Ordering::Less
+        } else {
+            Ordering::Greater
+        }
+    }
+    fn min(self, other: Self) -> Self
+    where
+        Self: Sized,
+    {
+        match self.cmp(&other) {
+            Ordering::Greater => other,
+            Ordering::Less => self,
+            Ordering::Equal => self,
+        }
+    }
+    fn max(self, other: Self) -> Self
+    where
+        Self: Sized,
+    {
+        match self.cmp(&other) {
+            Ordering::Greater => self,
+            Ordering::Less => other,
+            Ordering::Equal => self,
+        }
+    }
+}
+impl<'a> PartialOrd for TokenBasedVol<'a> {
+    fn ge(&self, other: &Self) -> bool {
+        !matches!(self.cmp(other), Ordering::Less)
+    }
+    fn le(&self, other: &Self) -> bool {
+        !matches!(self.cmp(other), Ordering::Greater)
+    }
+    fn gt(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Greater)
+    }
+    fn lt(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Less)
+    }
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::types::common::{
+        BigInt, Bytes, Order, Orderbook, TradeEvent, TradeStructPartialOrder,
+        TradeVaultBalanceChange, Transaction, Vault, VaultBalanceChangeVault,
+    };
+    use alloy::primitives::{Address, B256};
+    use std::str::FromStr;
+
+    #[test]
+    fn test_get_pairs_ratio() {
```

Comment:
test unhappy path?

Reply:
there is no unhappy path for this fn, it returns a hashmap